### ISSUE TITLE
Release v3

### ISSUE THREAD

**Issue by sosukesuzuki** [prettier/prettier#13142]
[next branch](https://github.com/prettier/prettier/pull/9583)

- [x] Support ESM plugins (#13201)
- [x] Support ESM config files (#13130)
- [x] Change `trailingComma` option default value to `all` (#11479)
- [x] Don't insert spaces in Markdown ( #11597 )
- [x] Remove `prettier.doc.builders.concat` (#13203)
- [ ] ~Update `print` function signature to not accept `path` (need investigate)~(Removed since this will break `path.map(print)`)
- [x] Print HTML5 doctype in lowercase? (need discussion #7391)
- [x] Change `--loglevel` flag to `--log-level` (#13204)
- [ ] ~Support MDX 2? (#12209)~ Postponed
- [x] `exports` field (#13530)
- [ ] Plugins & tools support for v3 (#13616)

***

- prepare
  - [x] Complete [milestone items](https://github.com/prettier/prettier/milestones)
  - [x] Run on other projects to check for regressions ( https://github.com/prettier/prettier-regression-testing/issues/348 )
  - [x] Write release notes (#13436)
    - use `node scripts/draft-blog-post.mjs` to generate a blog post draft (a file named <code>*YYYY*-*MM*-*DD*-*version*.md</code> in [`website/blog`](https://github.com/prettier/prettier/tree/master/website/blog)) from `changelog_unreleased`
    - [prs-merged-since](https://npm.im/prs-merged-since) ([readme](https://github.com/suchipi/prs-merged-since#prs-merged-since)) may be useful
- publish
  - [x] Run [release script](https://github.com/prettier/prettier/tree/main/scripts/release)
- postpublish
  - [ ] Merge release notes
  - [ ] Wait for release notes to deploy
  - [ ] Clean `changelog_unreleased`
    - use `node scripts/clean-changelog-unreleased.mjs`
  - [ ] Tweet from [`@PrettierCode`](https://twitter.com/PrettierCode) account via TweetDeck
  - [ ] Close milestone
  - [ ] Send schema PR generated by `node scripts/generate-schema.mjs` to [SchemaStore](https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/prettierrc.json) if we added a new option.
  - [ ] Update the [release checklist](https://github.com/prettier/prettier/wiki/Release-Checklist) with possible changes]
  - [ ] deprecates `@types/prettier`

Then monitor [Twitter](https://twitter.com/search?q=%40PrettierCode&src=typed_query&f=live)/[GitHub](https://github.com/prettier/prettier/issues) for any immediate issues 😄

**Comment by kachkaev** [comment]
Switching to MDX2 is also breaking and is due. We should probably include https://github.com/prettier/prettier/issues/12209 too.

**Comment by fisker** [comment]
We can include it if someone can work on it. (I don't use MDX, not interested)

**Comment by sosukesuzuki** [comment]
> Switching to MDX2 is also breaking and is due. We should probably include https://github.com/prettier/prettier/issues/12209 too.

I had worked on it before and will try again but it was very difficult...

**Comment by chrisbobbe** [comment]
Please consider adding PR https://github.com/prettier/prettier/issues/7111 ("JS: Break lines before binary operators (fixes #3806)") to the list. https://github.com/prettier/prettier/issues/3806#issuecomment-1104635354 is a nice status summary for this long-standing issue.

**Comment by ExE-Boss** [comment]
The following should probably also be included:
- <https://github.com/prettier/prettier/issues/7475>

**Comment by JounQin** [comment]
Can we rename `--vue-indent-script-and-style` to `--indent-script-and-style` for all SFC files like `svelte`?

[prettier-plugin-svelte](https://github.com/sveltejs/prettier-plugin-svelte#svelte-indent-script-and-style) has to have a new `--svelte-indent-script-and-style` option, but they could be unified, actually.

**Comment by fisker** [comment]
I don't think we are going to make more changes to the plugin system, (except #10156).

I think we are ready to release an alpha version for plugins to test.

**Comment by xsrvmy** [comment]
Other problems: some phrases like U盘 is sometimes typed without space (although fullwidth can be used here), and sometimes dates are typeset without spaces.
And of course prettier should not format in a way that affects output.

**Comment by JounQin** [comment]
> Other problems: some phrases like U盘 is sometimes typed without space (although fullwidth can be used here), and sometimes dates are typeset without spaces. And of course prettier should not format in a way that affects output.

See #11597

**Comment by thorn0** [comment]
Why not drop Node 14?

**Comment by sosukesuzuki** [comment]
Node.js 14 is maintenance LTS until April 30, 2023, so we need to support ithttps://github.com/nodejs/Release#release-schedule

**Comment by thorn0** [comment]
What if we drop it only for direct-from-github installs?

**Comment by fisker** [comment]
Sure we can.

**Comment by sosukesuzuki** [comment]
We can.

**Comment by thorn0** [comment]
How about completely removing support for Flow comment types? [They're a mess and can't be properly supported anyway.](https://github.com/facebook/flow/issues/8355)

Related issues: https://github.com/prettier/prettier/issues?q=is%3Aopen+is%3Aissue+label%3A%22area%3Aflow+comment+types%22

**Comment by sosukesuzuki** [comment]
> How about completely removing support for Flow comment types? They're a mess and can't be properly supported anyway. See https://github.com/facebook/flow/issues/8355

I agree with you.

**Comment by rassie** [comment]
Probably late to the party, but I'd *really* appreciate if something could be done about https://github.com/prettier/prettier/issues/6921. Pretty much anything to ease the pain.

**Comment by dummdidumm** [comment]
Would #7639 be something that's relevant to a major release (because it would require a breaking change to implement properly), and if yes, would you consider it for v3?

**Comment by fisker** [comment]
> Would #7639 be something that's relevant to a major release (because it would require a breaking change to implement properly), and if yes, would you consider it for v3?

Absolutely, but we need rethink about the API.


> because it would require a breaking change

I don't think add range format support to plugin is a breaking change, it's a feature. I think we can add it anytime we are ready.

**Comment by nickmccurdy** [comment]
Are we still going to consider implementing other issues in the [milestone](https://github.com/prettier/prettier/milestone/14) before Prettier 3 is stable? I'm especially interested in #7475 for accessibility.

**Comment by BrianHVB** [comment]
@sosukesuzuki - Do you have concerns about adding #3806 (PR #7111) to the checklist? This change has wide support, has been pending for over 3 years, was originally part of the 2.0 milestone, and is currently on the 3.0 milestone.

**Comment by Divlo** [comment]
Is it possible to take a look at: https://github.com/prettier/prettier/issues/14115, please? :smile:

**Comment by connorjs** [comment]
> Node.js 14 is maintenance LTS until April 30, 2023, so we need to support it https://github.com/nodejs/Release#release-schedule

Do we expect Prettier 3 to land before April 30th? (~45 days from now)

(I was reading the sample blog post and also was surprised to see 14 support staying. This is also a backwards way to get a pulse check on Prettier 3 release timing 😉.)

**Comment by fisker** [comment]
It seems we don't need features that only works for Node.js v16, so it's unnecessary to drop support for v14.

**Comment by connorjs** [comment]
Thanks for the answer! The main thing for a library (aside from language features) is full `exports` support IIRC, which is also in a later minor of 14. (v14.13 per <https://nodejs.org/api/packages.html#exports>), if y'all have any interest in using that.

**Comment by sosukesuzuki** [comment]
> It seems we don't need features that only works for Node.js v16, so it's unnecessary to drop support for v14.

@fisker Node.js 16 supports `Object.hasOwn`, `String.prototype.replaceAll`, `Array.prototype.at` so if we drop Node.js 14, we can remove all transforms from v3 build steps?

**Comment by fisker** [comment]
Don't know , it's just 30k https://github.com/prettier/prettier/actions/runs/4519735825/jobs/7960389853?pr=14589#step:4:268 .

I hope to keep it, unless you want use other features.

For example https://github.com/prettier/prettier/pull/13740, if it works in Node.js 16, I'll be happy to drop 14. Unfortunately it requires Node.js 18.

**Comment by fisker** [comment]
My most excited feature that works in 16 is `Intl.Segmenter`, but I can't find a place to use in this repo.

**Comment by sosukesuzuki** [comment]
> Don't know , it's just 30k

We would like to get rid of our own transform mechanism if possible. Adding a new layer means adding a new source of bugs. Of course, the transform step you created that allows the use of new JS features is great, but if we can use those features without it, I think that would be better.

> I hope to keep it, unless you want use other features.

Can you tell me why you want to keep supporting Node.js 14? I think the narrower the range of versions that should be supported, the easier it would be to maintain.

**Comment by fisker** [comment]
> Can you tell me why you want to keep supporting Node.js 14? 

Quote from https://github.com/prettier/prettier/issues/10157#issuecomment-768222172

> We must not forget about our users, even if they are in legacy world

**Comment by fisker** [comment]
Let's think this way,

Why drop Node.js v10? -- We want native ESM support.
Why drop Node.js v12? -- We can use TLA(top level await), and it's not possible to shim.
Why drop Node.js v14? -- ___?

**Comment by sosukesuzuki** [comment]
> We must not forget about our users, even if they are in legacy world

Ah, yes, this may be an important perspective.

> Why drop Node.js v12? -- We can use TLA(top level await), and it's not possible to shim.

Does our application code need TLA? (It is used for only development scripts?)
I thought [we dropped because just it had reached EoL](https://github.com/prettier/prettier/issues/10157#issuecomment-1185483173).

In this perspective, perhaps we should continue to support v12.

**Comment by fisker** [comment]
> Does our application code need TLA? 

We don't need for production, I said we _can_,  we can consider dropping 12 improves our develop experience. I can't see useful features we need to drop 14.

> In this perspective, perhaps we should continue to support v12.

Maybe we can, but should be hard to test. Besides it's already 10 months, not sure worth the effort. I'll be happy to see it supportted, if someone want work on it, and not causes pain.

**Comment by fisker** [comment]
One question about `.at()`. 

You may missed https://github.com/prettier/prettier/pull/14396#issuecomment-1439316248

Do you think we should we force to use` [.length - _]` transform in all version?

**Comment by sosukesuzuki** [comment]
> We don't need for production, I said we can, we can consider dropping 12 improves our develop experience. I can't see useful features we need to drop 14. If you can find some, let me know.

I got it.

> Maybe we can, but should be hard to test. Besides it's already 10 months, not sure worth the effort. I'll be happy to see it supportted, if someone want work on it, and not causes pain.

I got it.

Let's make the policy for avoiding to repeat similar discussions in the future. It is probably easier to understand if written in code than in English, so I will write it in pseudo code. What do you think about in following policy?

```
if (
  (VERSION reached EoL) && (
    (We need to use a new feature that is not available in VERSION) ||
    (It has been a year since VERSION reached EoL)
  )
) {
  drop(VERSION)
}
```

> Do you think we should we force to use [.length - _] transform in all versions?

I missed it,  I think we should leave transform until `at` is faster. Thanks.

**Comment by fisker** [comment]
> What do you think about in following policy?

Sounds good.

---

On 16 and 18, we are calling `.at` even with the transform.

https://github.com/prettier/prettier/blob/4edb68ac50a847bef0d2968e1a1e4de643add7b0/scripts/build/shims/at.js#L13-L15

I mean, we should use property access even `.at` is available.

**Comment by sosukesuzuki** [comment]
> I mean, we should use property access even .at is available.

I see. I think we should use property access even `.at` is available.

**Comment by fisker** [comment]
Reply to https://github.com/prettier/prettier/issues/14527#issuecomment-1483882763

I'll check if I have changes want to make in v3, will let you know on monday. Let's set the date to **April 1**?

**Comment by fisker** [comment]
> I think we should use property access even .at is available.

Would you like to work on it?

**Comment by sosukesuzuki** [comment]
> I'll check if I have changes want to make in v3, will let you know on monday. Let's set the date to April 1?

:+1:

> Would you like to work on it?

Yes.

**Comment by sosukesuzuki** [comment]
FYI I've created to new wiki page  https://github.com/prettier/prettier/wiki/The-policy-to-drop-Node.js-version

**Comment by martinszeltins** [comment]
What about rewriting Prettier in TypeScript for version 3.0? Why is Prettier not written in TypeScript in the first place?

**Comment by kachkaev** [comment]
@martinszeltins v3 makes a step in this direction by shipping with typings (https://github.com/prettier/prettier/pull/14212), thus replacing [@types/prettier](https://www.npmjs.com/package/@types/prettier). Generating those types from source would be even better but that’s currently out of scope. Feel free to help when v3 is released!

**Comment by kachkaev** [comment]
The latest v3 alpha to date is [3.0.0-alpha.6](https://www.npmjs.com/package/prettier/v/3.0.0-alpha.6), which was released on Feb 3 (more than three months ago). I don’t remember if there are any risky changes merged into `main` after that, but if anyone thinks we should release another `alpha`, happy to test it!

**Comment by Mister-Hope** [comment]
> Node.js 14 is maintenance LTS until April 30, 2023, so we need to support it https://github.com/nodejs/Release#release-schedule

So can we relook into dropping Node14 support? I don't think prettier will touch stable before this date?

**Comment by fisker** [comment]
We discussed Node.js support recently https://github.com/prettier/prettier/wiki/The-policy-to-drop-Node.js-version

I don't understand why anyone except maintainers want support less versions ... Just curious.

**Comment by Mister-Hope** [comment]
> don't understand why anyone except maintainers want support less versions ... Just curious.

As long as there is any performance improvement by dropping node14 (even <1%) then I am positive towards this.

**Comment by kachkaev** [comment]
It takes time for the community to migrate away from old Node versions, even after their end of life. I’m up for dropping Node 14 in Prettier v4 (not 3), but now, even if release on May 1, it would be a bit too early.

**Comment by re-thc** [comment]
> I don't understand why anyone except maintainers want support less versions ... Just curious.

To move the world forward. Otherwise we'd always still be supporting IE too.

**Comment by fisker** [comment]
Just want to brief here, we are blocked by VSCode support. https://github.com/prettier/prettier-vscode/pull/2947

**Comment by ntkoopman** [comment]
There is still no conclusion to https://github.com/prettier/prettier/issues/3806. I feel this long standing issue should be addressed before v3, even if it's closing it as "won't fix" so we can move on.

**Comment by sosukesuzuki** [comment]
we'll collect feedback from users via [experimental flags](https://github.com/prettier/prettier/issues/14527) after v3 is released. So there is no need to draw any conclusions about it yet.

**Comment by Janther** [comment]
Why is the latest release [3.0.0-alpha.6](https://github.com/prettier/prettier/releases/tag/3.0.0-alpha.6) on Github while on npm is [3.0.0-alpha.12](https://www.npmjs.com/package/prettier/v/3.0.0-alpha.12).
There used to be a `next` branch as well.

**Comment by latin-1** [comment]
It appears that v3 no longer autoloads plugins. Shouldn't this be mentioned in the release notes?

**Comment by sosukesuzuki** [comment]
released

**Comment by karlhorky** [comment]
Thanks to everyone involved in Prettier v3!! 🙌 🎉 🚀
